File: ALG/C++/bin_search.cpp
#include <iostream>


class Bin_search{
    public:
        Bin_search(int elem, int size = 10) : size(size), elem(elem), arr(new int[size]){
            create_arr();
            print();
            std::cout << elem << " has index " << search(this->arr, this->elem, 0, this->size) << std::endl;
        };

    protected:
        void create_arr(){
            for (int i = 0; i < this->size; i++){
                this->arr[i] = i+1;
            }    
        };
        void print(){
            for (int i = 0; i < this->size; i++){
                std::cout << this->arr[i] << " ";
            }
            std::cout << std::endl;
        };
        int search(int *arr, int elem, int start, int end){
            if (start <= end){
                int mid = (start + end) / 2;
                if (arr[mid] == elem){
                    return mid;
                }
                if (arr[mid] > elem){
                    return search(arr, elem, start, mid - 1);
                }
                if (arr[mid] < elem){
                    return search(arr, elem, mid + 1, end);
                }
            } else{
                return -1;
            }
        };

    private:
        int size;
        int *arr;
        int elem;
};


int main(){
    Bin_search bs(5);
    return 0;
}

File: ALG/Python/block1/bin_search.py
def bin_search(arr, elem, start, end):
    if start <= end:
        mid = (start + end) // 2
        if arr[mid] == elem:
            return mid
        if arr[mid] > elem:
            return bin_search(arr, elem, start, mid - 1)
        if arr[mid] < elem:
            return bin_search(arr, elem, mid + 1, end)
    else:
        return -1


def tests():
    assert bin_search(lst, 5, 0, len(lst)) == 4
    assert bin_search(lst, 3, 0, len(lst)) == 2
    assert bin_search(lst, 7, 0, len(lst)) == 6


lst = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
if __name__ == "__main__":
    tests()
